mono/metadata/file-mmap-windows.c

```diff
@@ -1,71 +1,413 @@
+		if (!existed && mode == FILE_MODE_CREATE_NEW && *capacity == 0) {
+			*error = CAPACITY_SMALLER_THAN_FILE_SIZE;
+			goto done;
+		}
+		hFile = CreateFileW (w_path, get_file_access (access), FILE_SHARE_READ, NULL, mode, FILE_ATTRIBUTE_NORMAL, NULL);
+		if (hFile == INVALID_HANDLE_VALUE) {
+			*error = convert_win32_error (GetLastError (), COULD_NOT_OPEN);
+			goto done;
+		}
+		delete_on_error = !existed;
+	}
 
-int
-mono_mmap_map (void *handle, gint64 offset, gint64 *size, int access, void **mmap_handle, void **base_address)
+	result = open_handle (hFile, mapName, mode, capacity, access, options, error);
+
+done:
+	if (!result && delete_on_error)
+		DeleteFileW (w_path);
+	if (w_path)
+		g_free (w_path);
+
+	return result;
+}
+
+void *mono_mmap_open_handle (void *handle, MonoString *mapName, gint64 *capacity, int access, int options, int *error)
 {
-	g_error ("No windows backend");
+	g_assert (handle != NULL);
+
+	return open_handle (handle, mapName, FILE_MODE_OPEN, capacity, access, options, error);
+}
+
+void mono_mmap_close (void *mmap_handle)
+{
+	g_assert (mmap_handle);
+	CloseHandle ((HANDLE) mmap_handle);
+}
+
+void mono_mmap_configure_inheritability (void *mmap_handle, gboolean inheritability)
+{
+	g_assert (mmap_handle);
+	if (!SetHandleInformation ((HANDLE) mmap_handle, HANDLE_FLAG_INHERIT, inheritability ? HANDLE_FLAG_INHERIT : 0)) {
+		g_error ("mono_mmap_configure_inheritability: SetHandleInformation failed with error %d!", GetLastError ());
+	}
+}
+
+void mono_mmap_flush (void *mmap_handle)
+{
+	g_assert (mmap_handle);
+	MmapInstance *h = (MmapInstance *)mmap_handle;
+
+	if (FlushViewOfFile (h->address, h->length))
+		return;
+
+	// This replicates how CoreFX does MemoryMappedView.Flush ().
+
+	// It is a known issue within the NTFS transaction log system that
+	// causes FlushViewOfFile to intermittently fail with ERROR_LOCK_VIOLATION
+	// As a workaround, we catch this particular error and retry the flush operation 
+	// a few milliseconds later. If it does not work, we give it a few more tries with
+	// increasing intervals. Eventually, however, we need to give up. In ad-hoc tests
+	// this strategy successfully flushed the view after no more than 3 retries.
+
+	if (GetLastError () != ERROR_LOCK_VIOLATION)
+		// TODO: Propagate error to caller
+		return;
+
+	for (int w = 0; w < MAX_FLUSH_WAIT; w++) {
+		int pause = (1 << w);  // MaxFlushRetries should never be over 30
+		mono_thread_info_sleep (pause, NULL);
+
+		for (int r = 0; r < MAX_FLUSH_RETIRES_PER_WAIT; r++) {
+			if (FlushViewOfFile (h->address, h->length))
+				return;
+
+			if (GetLastError () != ERROR_LOCK_VIOLATION)
+				// TODO: Propagate error to caller
+				return;
+
+			mono_thread_info_yield ();
+		}
+	}
+
+	// We got to here, so there was no success:
+	// TODO: Propagate error to caller
+}
+
+int mono_mmap_map (void *handle, gint64 offset, gint64 *size, int access, void **mmap_handle, void **base_address)
```

Comment:
This is an icall which uses `long` in C#. The icall on the C# side is called by public APIs which also use `long` for size. The size is validated on the managed size to be positive so we could change the internal C# methods to take `ulong`. But I'm not sure it's worth the hassle to change the the icall signatures.